File: ios/Sources/MaintenanceFieldCore/LoginStateMachine.swift
import Foundation
import MaintenanceAPIClient

public enum DeviceRegistrationRetryStatus: String, Equatable, Sendable {
    case readyForReplay
}

public struct DeviceRegistrationRetry: Equatable, Sendable {
    public let deviceID: String
    public let platform: Components.Schemas.DevicePlatform
    public let appVersion: String
    public let pushToken: String?
    public let status: DeviceRegistrationRetryStatus
    public let messageKey: String
    public let lastErrorClass: String

    public init(
        deviceID: String,
        platform: Components.Schemas.DevicePlatform,
        appVersion: String,
        pushToken: String?,
        status: DeviceRegistrationRetryStatus,
        messageKey: String,
        lastErrorClass: String
    ) {
        self.deviceID = deviceID
        self.platform = platform
        self.appVersion = appVersion
        self.pushToken = pushToken
        self.status = status
        self.messageKey = messageKey
        self.lastErrorClass = lastErrorClass
    }
}

public enum DeviceRegistrationStatus: Equatable, Sendable {
    case registered
    case retryPending(DeviceRegistrationRetry)
}

public enum LoginState: Equatable, Sendable {
    case signedOut(messageKey: String? = nil)
    case awaitingPasskey(userID: Components.Schemas.Uuid, ceremonyID: Components.Schemas.Uuid, challengeJSON: String)
    case registeringDevice(
        accessToken: String,
        refreshToken: String,
        deviceID: String,
        platform: Components.Schemas.DevicePlatform,
        appVersion: String
    )
    case authenticated(
        accessToken: String,
        refreshToken: String,
        deviceRegistration: DeviceRegistrationStatus = .registered,
        messageKey: String? = nil
    )
}

public enum LoginEvent: Equatable, Sendable {
    case loginChallengeReceived(
        userID: Components.Schemas.Uuid,
        ceremonyID: Components.Schemas.Uuid,
        challengeJSON: String,
        expiresAt: Date
    )
    case passkeyVerified(accessToken: String, refreshToken: String, deviceID: String, appVersion: String)
    case deviceRegistered(serverDeviceID: Components.Schemas.Uuid)
    case deviceRegistrationFailed(messageKey: String, lastErrorClass: String)
    case failed(messageKey: String)
}

public struct LoginStateMachine: Sendable {
    public init() {}

    public func reduce(_ state: LoginState, _ event: LoginEvent) -> LoginState {
        switch event {
        case let .loginChallengeReceived(userID, ceremonyID, challengeJSON, _):
            return .awaitingPasskey(userID: userID, ceremonyID: ceremonyID, challengeJSON: challengeJSON)

        case let .passkeyVerified(accessToken, refreshToken, deviceID, appVersion):
            switch state {
            case .awaitingPasskey:
                return .registeringDevice(
                    accessToken: accessToken,
                    refreshToken: refreshToken,
                    deviceID: deviceID,
                    platform: .ios,
                    appVersion: appVersion
                )
            default:
                return .signedOut(messageKey: "login_failed")
            }

        case .deviceRegistered:
            switch state {
            case let .registeringDevice(accessToken, refreshToken, _, _, _):
                return .authenticated(accessToken: accessToken, refreshToken: refreshToken)
            default:
                return .signedOut(messageKey: "login_failed")
            }

        case let .deviceRegistrationFailed(messageKey, lastErrorClass):
            switch state {
            case let .registeringDevice(accessToken, refreshToken, deviceID, platform, appVersion):
                return .authenticated(
                    accessToken: accessToken,
                    refreshToken: refreshToken,
                    deviceRegistration: .retryPending(
                        DeviceRegistrationRetry(
                            deviceID: deviceID,
                            platform: platform,
                            appVersion: appVersion,
                            pushToken: nil,
                            status: .readyForReplay,
                            messageKey: messageKey,
                            lastErrorClass: lastErrorClass
                        )
                    )
                )
            default:
                return .signedOut(messageKey: "login_failed")
            }

        case let .failed(messageKey):
            return .signedOut(messageKey: messageKey)
        }
    }
}

File: android/app/src/main/kotlin/com/maintenance/field/auth/LoginStateMachine.kt
package com.maintenance.field.auth

import com.maintenance.api.client.model.DevicePlatform
import java.time.OffsetDateTime
import java.util.UUID

sealed interface LoginState {
    data class SignedOut(val messageKey: String? = null) : LoginState

    data class AwaitingPasskey(
        val userId: UUID,
        val ceremonyId: UUID,
        val challengeJson: String,
    ) : LoginState

    data class RegisteringDevice(
        val accessToken: String,
        val refreshToken: String,
        val deviceId: String,
        val platform: DevicePlatform,
        val appVersion: String,
    ) : LoginState

    data class Authenticated(
        val accessToken: String,
        val refreshToken: String,
        val deviceRegistration: DeviceRegistrationState = DeviceRegistrationState.Registered,
    ) : LoginState
}

const val DEVICE_REGISTRATION_RETRY_PENDING_MESSAGE_KEY = "device_registration_retry_pending"

sealed interface DeviceRegistrationState {
    data object Registered : DeviceRegistrationState

    data class RetryPending(
        val deviceId: String,
        val platform: DevicePlatform,
        val appVersion: String,
        val lastErrorClass: String,
        val messageKey: String = DEVICE_REGISTRATION_RETRY_PENDING_MESSAGE_KEY,
    ) : DeviceRegistrationState
}

sealed interface LoginEvent {
    data class LoginChallengeReceived(
        val userId: UUID,
        val ceremonyId: UUID,
        val challengeJson: String,
        val expiresAt: OffsetDateTime,
    ) : LoginEvent

    data class PasskeyVerified(
        val accessToken: String,
        val refreshToken: String,
        val deviceId: String,
        val appVersion: String,
    ) : LoginEvent

    data class DeviceRegistered(val serverDeviceId: UUID) : LoginEvent

    data class DeviceRegistrationFailed(
        val lastErrorClass: String,
        val messageKey: String = DEVICE_REGISTRATION_RETRY_PENDING_MESSAGE_KEY,
    ) : LoginEvent

    data class Failed(val messageKey: String) : LoginEvent
}

class LoginStateMachine {
    fun reduce(state: LoginState, event: LoginEvent): LoginState = when (event) {
        is LoginEvent.LoginChallengeReceived -> LoginState.AwaitingPasskey(
            userId = event.userId,
            ceremonyId = event.ceremonyId,
            challengeJson = event.challengeJson,
        )
        is LoginEvent.PasskeyVerified -> LoginState.RegisteringDevice(
            accessToken = event.accessToken,
            refreshToken = event.refreshToken,
            deviceId = event.deviceId,
            platform = DevicePlatform.ANDROID,
            appVersion = event.appVersion,
        )
        is LoginEvent.DeviceRegistered -> when (state) {
            is LoginState.RegisteringDevice -> LoginState.Authenticated(
                accessToken = state.accessToken,
                refreshToken = state.refreshToken,
            )
            is LoginState.Authenticated -> state.copy(
                deviceRegistration = DeviceRegistrationState.Registered,
            )
            else -> LoginState.SignedOut(messageKey = "login_failed")
        }
        is LoginEvent.DeviceRegistrationFailed -> when (state) {
            is LoginState.RegisteringDevice -> LoginState.Authenticated(
                accessToken = state.accessToken,
                refreshToken = state.refreshToken,
                deviceRegistration = DeviceRegistrationState.RetryPending(
                    deviceId = state.deviceId,
                    platform = state.platform,
                    appVersion = state.appVersion,
                    lastErrorClass = event.lastErrorClass,
                    messageKey = event.messageKey,
                ),
            )
            is LoginState.Authenticated -> state
            else -> LoginState.SignedOut(messageKey = "login_failed")
        }
        is LoginEvent.Failed -> LoginState.SignedOut(messageKey = event.messageKey)
    }
}
